CellMeasures as cube metadata

marqh opened the pull request:
Cell Measures added as Cube metadata

support for load and save for CF netCDF

marqh commented:
hmmm, travis doesn't seem to have triggered

odd

pelson commented on lib/iris/coords.py:
```diff
@@ -1611,6 +1611,92 @@ def __hash__(self):
 
+class CellMeasure(AuxCoord):
+    """
+    CF Cell Measure
```
May as well delete or improve.

pp-mo commented:
Does this simply *replace* #1760 ??

marqh commented:
> Does this simply replace #1760 ??

yes it does, I consider #1760 obsolete now

dkillick commented on lib/iris/tests/unit/coords/test_CellMeasure.py:
```diff
@@ -0,0 +1,54 @@
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with Iris.  If not, see <http://www.gnu.org/licenses/>.
+"""Unit tests for the :class:`iris.coords.CellMeaasure` class."""
```
Typo: 'Meaasure'

dkillick commented on lib/iris/tests/unit/coords/test_CellMeasure.py:
```diff
@@ -0,0 +1,54 @@
+from iris.coord_systems import GeogCS
+
+
+class Test_CellMeasure(tests.IrisTest):
```
This test class is fairly sparse. At the least you should also have a test for setting the measure to something other than `area` or `volume`.

Given that you're only (currently) testing settable Cell Measures properties, you could replace `self.measure` below with a mocked cell measure object and confirm that the mocked object exhibits the appropriate behaviour without having to explicitly define a measure instance.

marqh replied:
> Given that you're only (currently) testing settable Cell Measures properties, you could replace self.measure below with a mocked cell measure object and confirm that the mocked object exhibits the appropriate behaviour without having to explicitly define a measure instance.

If I did this I would not be testing the __init__ for the class and added a bit of complexity to the test code.
I am not sure it adds value in this case.  Do you think this 'could' be done, or do you think it 'should' be done (and why)?
cheers

dkillick commented:
@marqh have you confirmed what impact adding cell measures to cubes will have on `merge` and `concatenate`?

marqh commented:
> @marqh have you confirmed what impact adding cell measures to cubes will have on merge and concatenate?

The reason for using the cube.metadata as a container for the _cell_measure_and_dims was explicitly to put merge and concatenate out of scope for this change

merge and concatenate will treat cell_measures as a static piece of metadata:
 * If all proto-cubes have exactly the same cell_measure collection, this will be propagated;
 * If any proto-cube has any difference in cell_measures then they will not be combined

pp-mo commented:
We'd really like to have this in release 1.9, if it's ready + agreed in time.
That's not a commitment, we'll see how it goes !

pp-mo commented on lib/iris/tests/integration/test_netcdf.py:
```diff
@@ -210,5 +210,28 @@ def test_lazy_mask_preserve_fill_value(self):
+    def setUp(self):
+        self.fname = tests.get_data_path(('NetCDF', 'ORCA2', 'votemper.nc'))
+
+    def test_load_raw(self):
```
I don't see the value of testing 'load_raw', 'load' and 'load_cube' separately here, as this just seems to be testing two things at once.
In fact, I think by design you can not have any merging when loading netcdf data from a *single* file, so there is no different behaviour to be expected from using "load_raw".  Neither is there any new behaviour there, specific to cell measures.
So I think I'd prefer this to just use "load_cube" throughout.

marqh replied:
I included this very deliberately.  I tripped over problems with merge and concatenate not preserving cell_measures

load_raw returned the cell measures, load_cube did not

so, i'd like to keep both of these to ensure that behaviour is maintained

pp-mo commented on lib/iris/tests/unit/experimental/regrid/test_regrid_weighted_curvilinear_to_rectilinear.py:
```diff
@@ -121,7 +121,7 @@ def _weighted_mean(self, points):
     def _expected_cube(self, data):
         cube = iris.cube.Cube(data)
-        cube.metadata = copy.deepcopy(self.src)
+        cube.metadata = copy.deepcopy(self.src.metadata)
```
This looks like it is a bug in the existing code.
Is that right ?

marqh replied:
i believe so, it surprised me

rhattersley replied:
It does look like a bug in the original ... so why didn't it fail before?

pp-mo commented:
Regardless of what this adds (and I still have more to check there), I think it has little effect on existing behaviour.

However, I did find this...
```
>>> cube = iris.load_cube(iris.tests.get_data_path(('NetCDF', 'ORCA2', 'votemper.nc')))
>>> [c.name() for c in cube.coords()]
[u'time', u'depth', u'latitude', u'longitude']
>>> cube.coords('cell_area')
[]
>>> cube.coord('cell_area')
CellMeasure(array([[  2.07163834e+09,   2.07163834e+09,   2.07163834e+09, ...,
          2.07163834e+09,   2.07163834e+09,   2.07163834e+09],
       [  2.21803803e+09,   2.21803803e+09,   2.21803803e+09, ...,
          2.21803803e+09,   2.21803803e+09,   2.21803803e+09],
       [  2.37452396e+09,   2.37452396e+09,   2.37452396e+09, ...,
          2.37452396e+09,   2.37452396e+09,   2.37452396e+09],
       ..., 
       [  1.55101938e+08,   1.20883644e+09,   1.91381624e+09, ...,
          2.48173924e+09,   1.91381624e+09,   1.20883644e+09],
       [  7.72599282e+07,   1.15935261e+09,   1.85087354e+09, ...,
          2.40648614e+09,   1.85087354e+09,   1.15935261e+09],
       [  3.51915941e+07,   1.12703331e+09,   1.81313209e+09, ...,
          2.36377103e+09,   1.81313209e+09,   1.12703331e+09]]), standard_name=u'cell_area', units=Unit('m2'), long_name=u'area of grid cell', var_name='areat')
>>> 
```

This definitely breaks my expectation of the relationship between ```cube.coords()``` and ```cube.coord()```.

When I work out just *how* it did that, I may find other implications.
But in any case, can we fix that ??

pp-mo commented on lib/iris/cube.py:
```diff
@@ -868,22 +892,27 @@ def _add_unique_aux_coord(self, coord, data_dims):
+        elif metadata.shape != (1,):
             raise ValueError('Missing data dimensions for multi-valued'
-                             ' coordinate {!r}'.format(coord.name()))
+                             ' {} {!r}'.format(mtype, coord.name()))
```
This fails when triggered, e.g. like this ...
```
>>> cube.add_aux_coord(AuxCoord([1,2]))   
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/h05/itpp/git/iris/iris_main/lib/iris/cube.py", line 883, in add_aux_coord
    self._add_unique_aux_coord(coord, data_dims)
  File "/home/h05/itpp/git/iris/iris_main/lib/iris/cube.py", line 915, in _add_unique_aux_coord
    'coordinate')
  File "/home/h05/itpp/git/iris/iris_main/lib/iris/cube.py", line 910, in _check_multi_dim_metadata
    ' {} {!r}'.format(mtype, coord.name()))
NameError: global name 'coord' is not defined
>>> 
```

pp-mo commented:
More leading from the above...
> I did find this...
  . . .
  \>\>\> cube.coords('cell_area')

You also get ...
```
>>> cube.remove_coord('cell_area')
>>> cube.remove_coord('cell_area')
```
when it should really ...
```
>>> cube.remove_coord('nonexistent_name')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/h05/itpp/git/iris/iris_main/lib/iris/cube.py", line 1044, in remove_coord
    coord = self.coord(coord)
  File "/home/h05/itpp/git/iris/iris_main/lib/iris/cube.py", line 1413, in coord
    raise iris.exceptions.CoordinateNotFoundError(msg)
iris.exceptions.CoordinateNotFoundError: 'Expected to find exactly 1  coordinate, but found none.'
>>> 
```
```

pp-mo commented:
**PLUS** 
More leading from the above...
> I did find this...
  . . .
  \>\>\> cube.coords('cell_area')

Due to the inheritance from AuxCoord, it is possible to add a CellMeasures ***as*** an AuxCoord ...
```
>>> cube.attributes={}
>>> print cube
sea_water_potential_temperature / (degC) (time: 1; depth: 31; -- : 148; -- : 180)
     Dimension coordinates:
          time                                x         -        -         -
          depth                               -         x        -         -
     Auxiliary coordinates:
          latitude                            -         -        x         x
          longitude                           -         -        x         x
     Cell Measures:
          cell_area                           -         -        x         x
     Cell methods:
          mean: time_counter
>>> measure = cube.cell_measures()[0]
>>> print measure
CellMeasure(array([[  2.07163834e+09,   2.07163834e+09,   2.07163834e+09, ...,
          2.07163834e+09,   2.07163834e+09,   2.07163834e+09],
       [  2.21803803e+09,   2.21803803e+09,   2.21803803e+09, ...,
          2.21803803e+09,   2.21803803e+09,   2.21803803e+09],
       [  2.37452396e+09,   2.37452396e+09,   2.37452396e+09, ...,
          2.37452396e+09,   2.37452396e+09,   2.37452396e+09],
       ..., 
       [  1.55101938e+08,   1.20883644e+09,   1.91381624e+09, ...,
          2.48173924e+09,   1.91381624e+09,   1.20883644e+09],
       [  7.72599282e+07,   1.15935261e+09,   1.85087354e+09, ...,
          2.40648614e+09,   1.85087354e+09,   1.15935261e+09],
       [  3.51915941e+07,   1.12703331e+09,   1.81313209e+09, ...,
          2.36377103e+09,   1.81313209e+09,   1.12703331e+09]]), standard_name=u'cell_area', units=Unit('m2'), long_name=u'area of grid cell', var_name='areat')
>>> cube.add_aux_coord(measure, (2,3))
>>> print cube
sea_water_potential_temperature / (degC) (time: 1; depth: 31; -- : 148; -- : 180)
     Dimension coordinates:
          time                                x         -        -         -
          depth                               -         x        -         -
     Auxiliary coordinates:
          cell_area                           -         -        x         x
          latitude                            -         -        x         x
          longitude                           -         -        x         x
     Cell Measures:
          cell_area                           -         -        x         x
     Cell methods:
          mean: time_counter
>>> 
>>> cube.coord('cell_area') == cube.cell_measure('cell_area')
True
>>> 
```

I can't work out why this would be a serious problem, but it seems like a doubtful idea.
I think coords and cell-measures really should exist in entirely separate spaces as far as cubes are concerned.

pp-mo commented on lib/iris/tests/unit/coords/test_CellMeasure.py:
```diff
@@ -0,0 +1,59 @@
+            self.measure.coord_system = GeogCS(6371229)
+
+    def test_invalid_measure(self):
+        msg = 'measure must be area or volume, not length'
```
It would be clearer if the names here had quotes - i.e. ```"area" or "volume", not "length"```.
You can handily use "{!r}" to print the arg in 'repr' form.

pp-mo commented on lib/iris/coords.py:
```diff
@@ -1611,6 +1611,97 @@ def __hash__(self):
 
+class CellMeasure(AuxCoord):
+    """
+    A CF Cell Measure, providing area or volume properties of a cell
```
As Iris itself makes no interpretation of the meaning of the 'measure' identity, I'm not sure we should be testing specifically for area/volume, or encoding any restriction on the name.
Likewise, here we can just say "providing a measure property of cells such as 'area' or 'volume'", and leave it at that.
As far as CF is concerned, a file can validly **contain** other measures, we just don't yet have any for which CF defines a specific meaning.

marqh replied:
> As far as CF is concerned, a file can validly contain other measures, we just don't yet have any for which CF defines a specific meaning.

not sure about this, CF appears to say that only area and volume are valid

pp-mo commented:
Inheritance from AuxCoord has given us some unwanted baggage :
```
>>> for x in dir(cube.cell_measure('cell_area')):
...   if x[0] != '_':
...     print x
... 
attributes
bounds
cell
cells
collapsed
contiguous_bounds
convert_units
coord_system
copy
dtype
from_coord
guess_bounds
has_bounds
intersect
is_compatible
is_contiguous
is_monotonic
long_name
measure
name
nbounds
ndim
nearest_neighbour_index
points
rename
shape
standard_name
units
var_name
xml_element
>>> 
```

Dividing that list up, I'd say these ones are ok ...
 * attributes
 * convert_units
 * dtype
 * copy
 * long_name
 * measure
 * name
 * points
 * nbounds
 * ndim
 * rename
 * shape
 * standard_name
 * units
 * var_name
 * xml_element

... these ones are disabled ...
 * bounds
 * collapsed
 * coord_system
 * guess_bounds

.. and these ones are perhaps questionable ...
 * cell
 * cells (works if 1-D)
 * contiguous_bounds  (works if 1D and monotonic)
 * from_coord  (returns a Coord, ***not*** a CellMeasure)
 * has_bounds
 * intersect
 * is_compatible
 * is_contiguous
 * is_monotonic
 * nearest_neighbour_index

marqh commented:
i've removed the cube.coord('name') returning a CellMeasure and implemented the functionality in cube.coord_dims instead.  perhaps this is better, given some of the comments so far

marqh commented:
> .. and these ones are perhaps questionable ...

I'm happy to disable any of these behaviours if it's helpful

>     cell
>     cells (works if 1-D)

'Return the single Cell instance which results from slicing the points/bounds with the given index.'
slicing is supported, so this seems valid to me

>     contiguous_bounds (works if 1D and monotonic)

an odd one, I assumed this would alter the coord and fail, but it just return an array.  Do you want this disabled?

>     from_coord (returns a Coord, not a CellMeasure)

nasty; this one should go

>     has_bounds

should give a sensible answer; i say keep it

>     intersect
>     is_compatible

Could always return false? This might be better than disabling, as this would cause intersect to always fail.

>     is_contiguous

should always be wrong.  disable, or force to return 'False'?

>     is_monotonic

should always be wrong.  disable, or force to return 'False'?

>     nearest_neighbour_index

'Returns the index of the cell nearest to the given point. Only works for one-dimensional coordinates.'

not valid; probably disabled?

rhattersley commented:
The relationship between a subclass and its parent class can be thought of as "is a". For example, "Dog" is-a "Mammal", so everything that is true for a mammal is also true for a dog. So if only *some* of the inherited attributes/methods are valid, then the subclass relationship itself is not valid.

rhattersley commented:
> We'd really like to have this in release 1.9, if it's ready + agreed in time.
> That's not a commitment, we'll see how it goes !

Agree. But so we can use the milestone for *blockers* I'm going to remove it from milestone.

marqh commented:
> The relationship between a subclass and its parent class can be thought of as "is a". For example, "Dog" is-a "Mammal", so everything that is true for a mammal is also true for a dog. So if only some of the inherited attributes/methods are valid, then the subclass relationship itself is not valid.

in my initial implementation the number of differences felt tiny, so I thought it a useful approach.  If this number of differences exist then the inheritance needs to be reconsidered.

I will assume that this is the case and look at a different inheritance pattern, unless people leap in to defend CellMeasure "is an" AuxCoord

marqh commented:
> Agree. But so we can use the milestone for blockers I'm going to remove it from milestone.

definitely, this is not a blocker, it doesn't belong on the milestone

marqh commented:
https://github.com/marqh/iris/commit/43875f5a797d2029af36702519dcd8aaa641e6b3
contains a rethinking of the inheritance pattern for CellMeasures and AuxCoords

Before I put in the extra effort on rewording all the documentation strings, reimplementing testing and otherwise fixing issues, please may I have a second opinion on whether this implementation has merit and might be a sensible improvement on the previous one?

many thanks
mark

pp-mo commented:
Some feedback that might just been helpful...
I analysed where the definitions of various properties were inherited from (at the previous version, i.e. https://github.com/marqh/iris/commit/8f2f2c4d9a9bb40582d115693b74914b0429126d).

I didn't do all the special/private ones, though...
```
By inheritance depth:
          CFVariableMixin   attributes
          CFVariableMixin   long_name
          CFVariableMixin   name
          CFVariableMixin   rename
          CFVariableMixin   standard_name
          CFVariableMixin   units
          CFVariableMixin   var_name
        Coord               cell
        Coord               cells
        Coord               collapsed
        Coord               contiguous_bounds
        Coord               convert_units
        Coord               copy
        Coord               dtype
        Coord               guess_bounds
        Coord               has_bounds
        Coord               intersect
        Coord               is_compatible
        Coord               is_contiguous
        Coord               is_monotonic
        Coord               nbounds
        Coord               ndim
        Coord               nearest_neighbour_index
        Coord               shape
        Coord               xml_element
      AuxCoord              from_coord
      AuxCoord              points
    CellMeasure             bounds
    CellMeasure             coord_system
    CellMeasure             measure

NOTES:
THE ESSENTIAL MISSING POINT IN THIS (the private/special we actually care about)
is __getitem__.
    * __getitem__ is defined at Coord level (though DimCoord extends it slightly)
        - we would want to replicate this, as for "points" + related.
    * 'points' is not concretely defined by Coord
        - DimCoord and AuxCoord implement it independently
    * AuxCoord supports a lazy array, while DimCoord does not
        (DimCoords are strictly 1-D and monotonic, so never "really big")

```

pp-mo commented:
A possible alternative approach ?

From a **practical** PoV, CellMeasures are indeed "almost" like a special kind of AuxCoord
 * without bounds
 * not appearing in the cube coords lists (attached elsewhere)

The problem is that many of the properties of a Coord (and/or AuxCoord)
are *not* required , desired or even sensible for CellMeasures, so we need to
"disable" them, which makes a bit of a mess.

It occurs to me that If the info were "hidden" in having an underlying AuxCoord implementation with 
a restricted interface via a wrapper object, would that fix the problem ???
 * a CellMeasures would "contain" an AuxCoord instead of inheriting from it
 * all necessary properties need to be reflected to the inner object
 * the constraints on the AuxCoord can be assumed by the wrapper, as it is in complete control

marqh commented:
thank you for sharing your idea @pp-mo 
i can see how this approach might work in theory, so I have tried.  however, in practice I got really lost in the details of the implementation and how to handle the contained object effectively.

I became worried about this approach, so I pulled back and attempted an implementation of CellMeasures which is independent of Coord all together.

I had to reimplement a few behaviours, but in most cases this resulted in a huge simplification of behaviour for the CellMeasure compared to similar behaviour for Coords and AuxCoords.
The result is an implementation which doesn't touch the current Coord implementation and is much lighter in changing the Cube, with predominantly additions.

There may be opportunity for neater methods of code sharing, but  I fear that the amount of change to the key classes will be large, with the opportunities for unintended consequences significant.

jkettleb commented:
Sorry to butt in on this one, and relatively late in the day.

Is there an implementation where the cell_measures are themselves cubes?   I'm sure there are pro's and con's to this - some of which I don't know as I don't know the internals of iris well enough. 

In the CMIP6 use case, that I think was one of the triggers for this, we will use the cell_measures as if they were a cube (we  need them to write them to a file - with their coordinates and meta data).  If the cell measures aren't a cube we'll probably need to implement a `cell_measures_as_cubes(source_cube)` function in the CMIP6 software.

There are obviously other uses of cell_measures though, apart from the CMIP6.  I can think of:

   1. use in an area/volume integral (like calculating the total heat content of the ocean, or an ocean basin)
   2. use in an area weighted regrid.

I think both these cases the use of cell_measures is much 'closer' to the data - so maybe there isn't such a strong case for keeping coordinates on the cell_measures in this case.

(So I think I'm not totally convinced myself - but I know we'll have more work to do our end if we don't have access to cell_measures as cubes - which means I think its worth asking)

For what its worth I definitely prefer the latest implementation more than the AuxCoords implementation - I have a mental block thinking of cell_measures as a coordinate.

rhattersley commented:
Has the on-going discussion in http://cf-trac.llnl.gov/trac/ticket/145 been considered in the design? In other words, will this PR leave the door open to the tweaks being suggested in that discussion?

jkettleb commented:
Thanks @rhattersley.   I think its still almost certain that CMIP6 will have `external_variables` for `cell_measures`, and this complicates things for reading those CMIP6 files.
 @marqh would know better - but I think you could accommodate that in what he has so I don't think we need to worry too much about that discussion on this pull request. 

 Our current need is more urgent than the CF discussion (and allowing us to get at the cell_measures from values NEMO/CICE files is needed whatever  the outcome of the discussion).

marqh commented:
> Has the on-going discussion in http://cf-trac.llnl.gov/trac/ticket/145 been considered in the design?

I agree with @jkettleb 
I think that if cf#145 is adopted then we might like to offer a little extra capability to enable cross file variable matching.  This may need a fair amount of user input, or it may be quite machineable, I'm not sure.

We'd probably limit it to load in the first instance

Either way, I think the core capability is pretty independent of cf#145. The concepts are the same, that mechanism allows for distributed storage

I would advocate handling CellMeasures as defined in CF1.5, then extending.

>  In other words, will this PR leave the door open to the tweaks being suggested in that discussion?

I believe that this PR will leave plenty of space in the doorway.  The 'add_cell_measure_and_dims' function gives us a lot of what we might want in this space.

marqh commented:
> Is there an implementation where the cell_measures are themselves cubes?

I thought about this, but I think the costs are large and the benefits small, in general.

Most usage of CellMeasures requires consistency of coordinates with the cube; as we don't share coordinate instances across cubes, making the cell measure a part of the cube seems to deliver this encapsulation with minimal effort.

This enables generic loading and saving of Cubes with netCDF

Implementing a 'make me a cube from the cell measure in this cube' capability in some library looks like a pretty achievable task, given this PR.
I think this could be put together without too much bother to meet the CMIP6 use case, if it is required.  It may well not be needed in Iris, or it may belong nicely there; we can see as that arises

marqh commented:
> There are obviously other uses of cell_measures though, apart from the CMIP6. I can think of:
> 
>     use in an area/volume integral (like calculating the total heat content of the ocean, or an ocean basin)
>     use in an area weighted regrid.

I haven't addressed these usage patterns automatically.  clearly passing in the cell measures as weights is a likely scenario

We can debate the benefits of making this automatic once an implementation is agreed upon, imho

jkettleb commented:
Thanks @marqh - I think we can work with your answer - we'll plan to support that function 'make me a cube from the cell_measure of this cube'.

No further comments on the interface etc from me - what other review is needed for merge?  (Sorry, but we are keen for this to go in).

pp-mo commented:
> Implementing a 'make me a cube from the cell measure in this cube' capability

In other contexts, I have more than once observed a need for creating a standalone cube from a coordinate.  If CellMeasures are implemented in the "obvious" way I think we should be able to create a generic Python function to do this with **either** CellMeasures or Coordinates from a cube.

> I have a mental block thinking of cell_measures as a coordinate

Nonetheless, I think there ***is*** a close association, which will try to state here :
 * An AuxCoord (CellMeasures) carries auxiliary values associated with a cube, with the following properties:
   * it is multidimensional array-like (*except*: cubes have data, coords have points + (optionally) bounds)
   * it's dimensions match some of the cube dimensions (with all the same coordinate associations)
   * it has the common properties of cube and coords ([CFVariableMixin class](https://github.com/SciTools/iris/blob/1ba7405ad86e5d8d2ea65b5f9a6510802775045d/lib/iris/_cube_coord_common.py#L86))
     * phenomenon identity encoded with standard_name / long_name / var_name
     * units
     * dtype, shape, ndims

A possibly useful viewpoint : *Basically, this is why we allow arithmetic between cubes and coordinates :  It makes perfect sense, as long as they have common dimensions.*

So by analogy, I think a duck-typing sort of conversion method can be written, to enable you to treat a coord (or CellMeasures) ***as*** a cube and thus handle it separately (e.g. to save).

In fact, it would make even *more* sense, IMHO, to extract an AuxCoord (or CellMeasure) from a cube "as" a separate cube, because then it will still have the associated DimCoords which reference the data, preventing it becoming "just an array".
At present, for example, whenever you call "add_aux_coord", you are asserting that the values are based "at" the appropriate dimension points, which is otherwise not a known fact.  Then if you extract a coordinate with cube.coord(), those associations are actually lost, but with a "coord_cube()" (or "cell_measure_cube()") method we would avoid that.

jkettleb commented:
Thanks @pp-mo.  Just for information - I don't think it should affect the completion of this pull request.  The reason I have a mental block over cell_measures being coordinates is I think of coordinates as telling where you are in some 'space'.   Whereas the cell_measures give you information on the properties of the space in that location (volume, area). 

We'd be happy, at this stage, to have the cell_measures exposed as the `CellMeasures` class as @marqh has them - we can turn it into a cube later.  What needs to happen to get sign off on this?  (sorry I know you have lots on - but we need access to the area's of the ocean grid for the CMIP6 work)

marqh commented:
I agree that a future enhancement would allow a CellMeasure and perhaps a Coordinate to be taken out of a cube and made its own cube.
I think this would be possible to do without too much more work, doing it by hand isn't that hard, it just takes some care

I don't think that should hold up this PR.  I think we can merge this, if it is suitable, then address 'toCube' later

pp-mo commented:
Still looking into this, and mulling over what it means.

I think that, since you fixed the behaviour of cube.coord(), the enhanced ```cube.__str__``` and/or ```cube.__repr__``` are no longer functioning :
```
>>> test_cube = Cube([1, 2, 3])
>>> test_cube.add_dim_coord(DimCoord([0, 1, 2], long_name='x'), (0,))
>>> print test_cube
unknown / (unknown)                 (x: 3)
     Dimension coordinates:
          x                           x
>>> 
>>> test_cube.add_cell_measure(
...     CellMeasure(data=np.array([0.1, 0.11, 0.14]),
...                 long_name='x_a',
...                 measure='area'),
...     (0,))
>>> 
>>> print test_cube
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/h05/itpp/git/iris/iris_main/lib/iris/cube.py", line 2087, in __str__
    return self.summary().encode(errors='replace')
  File "/home/h05/itpp/git/iris/iris_main/lib/iris/cube.py", line 1963, in summary
    vector_cell_measures, cube_header, max_line_offset)
  File "/home/h05/itpp/git/iris/iris_main/lib/iris/cube.py", line 1918, in vector_summary
    dims = self.coord_dims(coord)
  File "/home/h05/itpp/git/iris/iris_main/lib/iris/cube.py", line 1112, in coord_dims
    coord = self.coord(coord)
  File "/home/h05/itpp/git/iris/iris_main/lib/iris/cube.py", line 1424, in coord
    dim_coords=dim_coords)
  File "/home/h05/itpp/git/iris/iris_main/lib/iris/cube.py", line 1348, in coords
    defn = coord._as_defn()
AttributeError: 'CellMeasure' object has no attribute '_as_defn'
>>> 
```

pp-mo commented:
Broadly speaking, I think this needs more testing for various cube operations that could be affected. 

There is a particular problem with methods that produce "derived" cube results -- it's obviously okay if they *remove* all the cell measures (at least for now), but not if they don't and the result "should" have different measures.
There are a bunch of operations resulting in "extracted/partial" cubes (as opposed to derived ones), that may potentially be affected.  I made this list:
 *   __getitem__
 *   extract
 *   intersect
 *   regrid
 *   interpolate
 *   collapsed
 *   aggregated_by
 *   copy

This is not as complex as it potentially sounds
 * firstly, many operations are implemented through indeixing, which already works.
 * secondly, in many other cases all that needs doing is to ensure the results don't have the cell measures (at least for now)

pp-mo commented:
I think CellMeasure should also probably have a ```dtype``` property, analagous to Cube and Coord, (as with ```shape``` + ```ndim```)

pp-mo commented on lib/iris/cube.py:
```diff
@@ -62,7 +62,8 @@ class CubeMetadata(collections.namedtuple('CubeMetadata',
                                            'attributes',
-                                           'cell_methods'])):
+                                           'cell_methods',
+                                           'cell_measures_and_dims'])):
```
This seems quite neat, and it makes ```cube.copy()``` work correctly, but unfortunately, I don't think it is a good idea.

My mental model of ```cube.metadata``` is that it encapsulates only the **"non-dimensional"** information relating to the contained data.
(I do wish that was clearly stated somewhere, but that's another story).

Thus, at present, you can reasonably write code that copies metadata from one cube to another, e.g. like this ...
```
cube_z = cube_a * cube_b
if cube_a.units == '1':
    cube_z.metadata = cube_b.metadata
```
(N.B. I'm not saying that makes **good** sense, but it might in some known case).

Also, I think some existing code actually does this
 \- for example the interpolation scheme  [here](https://github.com/SciTools/iris/blob/b9e6c6d0d9b8b7216197212d989cc7cad568fb2f/lib/iris/analysis/_interpolation.py#L619)


At present, this means that 'metadata' is, basically, everything *except* coords + factories.
For consistency, I think this should *not* include cell_measures.

bjlittle replied:
@marqh This is a **really, really** fine point made by @pp-mo that I'd never considered ... and I agree with him in that cube metadata should be independent of cube dimensionality.

Perhaps there's no dodging the cube concatenate and cube merge bullets after all :cry: 

Are there any other opinions on whether dimensionality should be mixed into cube metadata, nudge @rhattersley @pelson et al.

rhattersley replied:
> This is a really, really fine point made by @pp-mo that I'd never considered ...

Agreed, and since he'd done an excellent job of making his point I didn't feel I had much to add. But since you've asked for an explicit opinion... well, it's implicit in my "Agreed", but to remove any remaining ambiguity ...

> ... 'metadata' ... should *not* include cell_measures

:+1:

> It probably makes perfectly good sense to simply remove cell measures from all interpolated outputs

Only if the interpolation relates to one or more dimensions spanned by the cell measure?

bjlittle commented on lib/iris/coords.py:
```diff
@@ -1628,6 +1628,195 @@ def __hash__(self):
+    def data(self, data):
+        # Set the data to a new array - as long as it's the same shape.
+        # If data are already defined for this coordinate,
+        if hasattr(self, '_data') and self._data is not None:
```
@marqh You can set the `data` to `None` but it breaks a whole bunch of things like `__str__()`, `ndim()` and `shape()`.

I don't think that it should be valid to do so ... coordinates at least enforce numeric data, shouldn't this be the case here?

marqh replied:
so, would you explicily ban the setting of self._data to None?

marqh replied:
i've raised a ValueError for the data setter with a payload of None.  _data needs to be able to be set to None, at least Transiently

bjlittle commented on lib/iris/coords.py:
```diff
@@ -1628,6 +1628,195 @@ def __hash__(self):
+    @data.setter
+    def data(self, data):
+        # Set the data to a new array - as long as it's the same shape.
+        # If data are already defined for this coordinate,
```
@marqh Hang-over from previous implementation? Or are we still thinking of cell measures as coordinates ...

bjlittle commented on lib/iris/coords.py:
```diff
@@ -1628,6 +1628,195 @@ def __hash__(self):
+
+    def __getitem__(self, key):
+        """
+        Returns a new Cell_Measure whose values are obtained by
```
@marqh Should be `CellMeasure` not `Cell_Measure`

bjlittle commented on lib/iris/coords.py:
```diff
@@ -1628,6 +1628,195 @@ def __hash__(self):
+
+        * data: A data array for the new cell_measure.
+                  This may be a different shape to the data of the
+                  cell_measure being copied.
```
@marqh Indent trauma ... and is it `cell measure`, `cell_measure` or `CellMeasure`

bjlittle commented on lib/iris/coords.py:
```diff
@@ -1628,6 +1628,195 @@ def __hash__(self):
+            # Explicitly not using the data property as we don't want the
+            # shape the new data to be constrained by the shape of
+            # self.data
+            new_cell_measure._data = None
```
@marqh Hmmm this doesn't smell right ... the implementation of how you're handling the `data` property is leaking out into this method and it shouldn't.

marqh replied:
this is to handle the implementation of the data setter, which enforces size consistency, and the __getitem__ which can change the size and shape

bjlittle commented on lib/iris/cube.py:
```diff
@@ -620,7 +621,8 @@ class Cube(CFVariableMixin):
                  cell_methods=None, dim_coords_and_dims=None,
-                 aux_coords_and_dims=None, aux_factories=None):
+                 aux_coords_and_dims=None, aux_factories=None,
+                 cell_measures_and_dims=None):
```
@marqh `cell_measures_and_dims` requires an associated entry in the `Kwargs` section of the doc-string.

bjlittle commented on lib/iris/cube.py:
```diff
@@ -1389,6 +1486,81 @@ def coord_system(self, spec=None):
+
+        """
+        name = None
+        coord = None
```
@marqh Not used

bjlittle commented on lib/iris/cube.py:
```diff
@@ -1054,6 +1132,25 @@ def coord_dims(self, coord):
+                   cm_ is cell_measure]
+
+        if not matches:
+            raise iris.exceptions.CoordinateNotFoundError(coord.name())
```
@marqh This exception `CoordinateNotFoundError` doesn't really apply now

bjlittle commented on lib/iris/cube.py:
```diff
@@ -1389,6 +1486,81 @@ def coord_system(self, spec=None):
+                bad_name = (cell_measure and cell_measure.name()) or ''
+            msg = 'Expected to find exactly 1 %s cell_measure, but found ' \
+                  'none.' % bad_name
+            raise iris.exceptions.CoordinateNotFoundError(msg)
```
@marqh As above, guess we should have `CellMeasureNotFoundError` exception

bjlittle commented on lib/iris/cube.py:
```diff
@@ -1054,6 +1132,25 @@ def coord_dims(self, coord):
+        # Search for existing cell measure (object) on the cube, faster lookup
+        # than equality - makes no functional difference.
+        matches = [dims for cm_, dims in self._cell_measures_and_dims if
+                   cm_ is cell_measure]
```
@marqh Are we sure that we really want to use `is` here ...

marqh replied:
no, but is consistent with 
cube.py:1121

bjlittle replied:
@marqh Okay, agreed ... but as a compromise I'd love for you to drop the appending of `_` to variable names, it's really not necessary here. Pretty please :pray:

marqh commented:
phew, i think that's concatenate and merge running usefully

@bjlittle what do you think?

marqh commented:
tests passing \o/

pp-mo commented:
!! Can't resist !!
Does this remind you of...

![Reminds me of this..](http://imgs.xkcd.com/comics/git_commit.png)

rhattersley commented:
> !! Can't resist !!

:grin: 

Perhaps time for a judicious squash. :+1: :wink:

marqh commented:
> Perhaps time for a judicious squash

indeed, i don't think i've typed
```
git rebase -i HEAD~19
```
before

rhattersley commented:
Interesting - I don't think I've ever typed `git rebase -i HEAD~...` - I always do `git rebase -i upstream/master` so I don't have to think about it.

bjlittle commented on lib/iris/coords.py:
```diff
@@ -160,7 +160,7 @@ class Cell(collections.namedtuple('Cell', ['point', 'bound'])):
     relative comparisons (lt, le, ..) are defined to be consistent with
     this interpretation. So for a given value `n` and Cell `cell`, only
-    one of the following can be true:
+    one f the following can be true:
```
@marqh Oops! `f` -> `of` ...

bjlittle commented on lib/iris/coords.py:
```diff
@@ -1628,6 +1628,214 @@ def __hash__(self):
+            The :class:`~cf_units.Unit` of the coordinate's values.
+            Can be a string, which will be converted to a Unit object.
+        * attributes
+            A dictionary containing other cf and user-defined attributes.
```
@marqh 'cf' -> 'CF'

bjlittle commented on lib/iris/coords.py:
```diff
@@ -1628,6 +1628,214 @@ def __hash__(self):
+    @property
+    def shape(self):
+        """The fundamental shape of the Cell Measure, expressed as a tuple."""
+        # Access the underlying _points attribute to avoid triggering
```
@marqh `... underlying _points attribute ...` -> `... underlying _data attribute ...`

bjlittle commented on lib/iris/coords.py:
```diff
@@ -1628,6 +1628,214 @@ def __hash__(self):
+        Return the number of dimensions of the cell measure.
+
+        """
+        return len(self.shape)
```
@marqh Minor point, but `ndim` is already available i.e. ```return self._data.ndim```

bjlittle commented on lib/iris/coords.py:
```diff
@@ -1628,6 +1628,214 @@ def __hash__(self):
+    def measure(self, measure):
+        if measure not in ['area', 'volume']:
+            raise ValueError('measure must be area or volume, '
+                             'not {}'.format(measure))
```
@marqh Just to reinforce that we're looking for strings here ...

```python
raise ValueError("Measure must be 'area' or 'volume', not {!r}".format(measure))
```

bjlittle commented on lib/iris/coords.py:
```diff
@@ -1628,6 +1628,214 @@ def __hash__(self):
+                    data = data[keys]
+                    if data.shape and min(data.shape) == 0:
+                        raise IndexError('Cannot index with zero length '
+                                         'slice.')
```
@marqh You could now slightly refactor ... note the negation

```python
        data = self._data
        if not all(is_full_slice(s) for s in full_slice):
            # Make indexing on the cube column based by using the
            # column_slices_generator (potentially requires slicing the
            # data multiple times).
            _, slice_gen = iris.util.column_slices_generator(full_slice,
                                                             self.ndim)
            for keys in slice_gen:
                if data is not None:
                    data = data[keys]
                    if data.shape and min(data.shape) == 0:
                        raise IndexError('Cannot index with zero length '
                                         'slice.')

bjlittle commented on lib/iris/coords.py:
```diff
@@ -1628,6 +1628,214 @@ def __hash__(self):
+
+    def _as_defn(self):
+        defn = CoordDefn(self.standard_name, self.long_name, self.var_name,
+                         self.units, self.attributes, self.measure)
```
@marqh This is a bit naughty! 

You're hijacking `CoordDefn` and all the actual parameters to formal parameters align apart from the actual parameter `self.measure` which is being mapped to the formal parameter `coord_system` in `CoordDefn` ... :bomb:

marqh replied:
eek, not intentionally, i'm fixing this

bjlittle commented on lib/iris/cube.py:
```diff
@@ -912,6 +927,38 @@ def add_aux_factory(self, aux_factory):
+
+        Args:
+
+        * cell_meaasure
```
@marqh `_meaasure` -> `_measure`

bjlittle commented on lib/iris/coords.py:
```diff
@@ -1628,6 +1628,214 @@ def __hash__(self):
+        if isinstance(other, CellMeasure):
+            eq = self._as_defn() == other._as_defn()
+            if eq:
+                eq = (self.data == other.data).all()
```
@marqh Gulp, this could be expensive ... this is more of an observation rather that asking you to do anything about it.

marqh replied:
indeed... but, if they're not the same, they're not the same

bjlittle commented on lib/iris/cube.py:
```diff
@@ -1389,6 +1472,80 @@ def coord_system(self, spec=None):
+        else:
+            cell_measure = name_or_cell_measure
+        cell_measures = []
+        for cm, dims in self._cell_measures_and_dims:
```
@marqh `dims` isn't used, so you could give the reader a hint to this by replacing `dims` with `_`

bjlittle commented on lib/iris/cube.py:
```diff
@@ -1389,6 +1472,80 @@ def coord_system(self, spec=None):
+        cell_measures = self.cell_measures(name_or_cell_measure)
+
+        if len(cell_measures) > 1:
+            msg = 'Expected to find exactly 1 cell_measure, but found %s. ' \
```
@marqh Oops! Change `%s` -> `{}`

bjlittle commented on lib/iris/cube.py:
```diff
@@ -1389,6 +1472,80 @@ def coord_system(self, spec=None):
+            msg = 'Expected to find exactly 1 cell_measure, but found %s. ' \
+                  'They were: {}.'.format(len(cell_measures),
+                                          ', '.join(cm.name() for cm in
+                                                    cell_measures))
```
@marqh You could break this out to aid readibility i.e.

```python
names = ', '.join('{!r}'.format(cm.name()) for cm in cell_measures)
msg = 'Expected to find ... They were: {}.'.format(len(cell_measures), names)
```
```

bjlittle commented on lib/iris/cube.py:
```diff
@@ -1389,6 +1472,80 @@ def coord_system(self, spec=None):
+            if isinstance(name_or_cell_measure, six.string_types):
+                bad_name = name
+            else:
+                bad_name = (cell_measure and cell_measure.name()) or ''
```
@marqh Oops! Usage should be `cell_measures` (plural) and **not** `cell_measure` i.e.

```python
bad_name = (cell_measures and cell_measures.name()) or ''
```

If you've not seen this, it suggests there ain't any test coverage :wink:

bjlittle commented on lib/iris/cube.py:
```diff
@@ -1050,6 +1114,25 @@ def coord_dims(self, coord):
+        matches = [dims for cm_, dims in self._cell_measures_and_dims if
+                   cm_ is cell_measure]
+
+        if not matches:
```
@marqh Your change is confused with the `coord_dims` method ...

This should should be ...

```python
if not matches:
    raise iris.exceptions.CellMethodNotFoundError(cell_measure.name())
```

bjlittle commented on lib/iris/cube.py:
```diff
@@ -1050,6 +1114,25 @@ def coord_dims(self, coord):
                        factories]
 
         if not matches:
+            raise iris.exceptions.CellMeasureNotFoundError(coord.name())
```
@marqh Oops! Something has gone wrong here, perhaps due to some rebasing issues?

Change this back to ...

```python
if not matches:
    raise iris.exceptions.CoordinateNotFoundError(coord.name())
```

Interesting how this change didn't cause any tests to fail ...

bjlittle commented on lib/iris/cube.py:
```diff
@@ -997,6 +1044,23 @@ def remove_coord(self, coord):
+        """
+        self._cell_measures_and_dims = [(cell_measure_, dim) for cell_measure_,
+                                        dim in self._cell_measures_and_dims
+                                        if cell_measure_ is not cell_measure]
```
@marqh `(cell_measure_, dim)` -> `(cm, dims)`

bjlittle replied:
@marqh Whoa! We need to be careful here ...

Your reconstructing the `Cube._cell_measures_and_dims` list to contain **tuples** i.e. `(cm, dims)`.

When you add a new cell measure with `Cube.add_cell_measure` it appends to `Cube._cell_measures_and_dims` with a **list** i.e. `[cm, dims]`

We need to be consistent in our approach, as I notice within `Cube.__getitem__` there's the call ...

```python
if self._cell_measures_and_dims != cube._cell_measures_and_dims:
```

bjlittle commented on lib/iris/cube.py:
```diff
@@ -912,6 +927,38 @@ def add_aux_factory(self, aux_factory):
+            raise ValueError('Duplicate cell_measures are not permitted.')
+        self._add_unique_cell_measure(cell_measure, data_dims)
+
+    def _add_unique_cell_measure(self, cell_measure, data_dims):
```
@marqh Is there a reason why you've created this private method ?

I can't find reference to its usage anywhere else in the code base other than from `Cube.add_cell_measure`.

Most likely it's existence is lingering from a previous implementation ... if so, could we blend this functionality directly into `Cube.add_cell_measure` ?

bjlittle commented on lib/iris/cube.py:
```diff
@@ -879,22 +889,27 @@ def _add_unique_aux_coord(self, coord, data_dims):
+        elif metadata.shape != (1,):
             raise ValueError('Missing data dimensions for multi-valued'
-                             ' coordinate {!r}'.format(coord.name()))
+                             ' {} {!r}'.format(mtype, metadata.name()))
```
@marqh I'd prefer that we didn't pass the `mtype` string into this method just to be used as part of a message in the exception.

You could replace the use of `mtype` with, for example `type(metadata).__name__` or `metadata.__class__.__name__`

marqh commented:
phew, that was a lot of little bits, impressive stamina @bjlittle 
thank you

these updates may represent all the actions, I tried to meet all of your change request comments

bjlittle commented on lib/iris/cube.py:
```diff
@@ -1679,6 +1836,10 @@ def summary(self, shorten=False, name_padding=35):
 
+            # cell measures
+            vector_cell_measures = [cm for cm in self.cell_measures()
+                                    if cm.shape != (1,)]
```
@marqh What if there is a cell measure in the cube with a shape of `(1,)` ... is this not dealt with anywhere ?

bjlittle commented on lib/iris/cube.py:
```diff
@@ -2001,6 +2175,15 @@ def __getitem__(self, keys):
             cube.add_aux_factory(factory.updated(coord_mapping))
 
+        # slice the cell measures, if changed and add them to the cube
+        if self._cell_measures_and_dims != cube._cell_measures_and_dims:
```
@marqh I don't see the point in this `if` statement, it may be left over from your previous implementation when cell measures were part of the cube metadata. Now, at this point in time, the new cube will always have an empty `_cell_measures_and_dims`.

bjlittle commented on lib/iris/exceptions.py:
```diff
@@ -40,6 +40,11 @@ class CoordinateNotFoundError(KeyError):
 
 
+class CellMeasureNotFoundError(KeyError):
+    """Raised when a search yields no coordinates."""
```
@marqh `coordinates` -> `cell measures`

bjlittle commented on lib/iris/_concatenate.py:
```diff
@@ -321,6 +321,7 @@ def __init__(self, cube):
         self.dim_metadata = []
         self.ndim = cube.ndim
         self.scalar_coords = []
+        self._cell_measures_and_dims = cube._cell_measures_and_dims
```
@marqh `_cell_measures_and_dims` -> `cell_measures_and_dims`

No need to make this private, the whole class is private, so let's follow the previous naming convention for the class attributes.

bjlittle commented on lib/iris/_concatenate.py:
```diff
@@ -464,6 +465,12 @@ def match(self, other, error_on_mismatch):
                                             self.data_type, other.data_type))
 
+        # Check _cell_measures_and_dims
+        if self._cell_measures_and_dims != other._cell_measures_and_dims:
```
@marqh Does the order of cell measures within `cell_measures_and_dims` matter?

bjlittle replied:
@marqh Update the doc-string to mention the cell measures are now considered part of the `_CubeSignature` matching criteria.

marqh replied:
> Does the order of cell measures within cell_measures_and_dims matter?

this is a tricky one.  I can see how we could decide that a reordered list is equivalent, but it would require more code to test for.  This is very much an edge case, it seems unlikely to occur.

do you want me to explicitly implement an unordered test?
if so, how might you suggest?

bjlittle replied:
@marqh From a CF perspective I could see that for cell **methods** the order might be important to the provenance and history of the data i.e. a mean then a variance is different to variance then a mean.

If we have a sequence of cell **measures ** for a cube and our understanding is that order doesn't matter (I'm kinda guided by your interpretation and expertise in all things CF here) then I'd be in favour of imposing a sort order which you could base on the `CellMeasure._as_defn()`.

If you agree with that as an approach, the next decision is at which point do you sort? I'd suggest it's imposed during `Cube.add_cell_measure` i.e. something along the lines of this (which combines `Cube.add_cell_measure` and `Cube._add_unique_cell_measure` ...

```python
class Cube(CFVariableMixin):
    ....
    def add_cell_measure(self, cell_measure, data_dims=None):
        if self.cell_measures(cell_measure):
            raise ValueError('Duplicate cell measures are not permitted.')
        data_dims = self._check_multi_dim_metadata(cell_measure, data_dims)
        cm_and_dims = self._cell_measures_and_dims
        cm_and_dims.append([cell_measure, data_dims])
        cm_and_dims.sort(key=lambda (cm, dims): (cm._as_defn(), dims))
```

bjlittle commented on lib/iris/_merge.py:
```diff
@@ -314,7 +314,7 @@ class _CoordSignature(namedtuple('CoordSignature',
 class _CubeSignature(namedtuple('CubeSignature',
                                 ['defn', 'data_shape', 'data_type',
-                                 'fill_value'])):
+                                 'fill_value', 'cell_measures_and_dims'])):
```
@marqh Please update the doc-string

bjlittle commented on lib/iris/_merge.py:
```diff
@@ -415,6 +415,9 @@ def match(self, other, error_on_mismatch):
                 msg = 'cube data fill_value differs: ' \
                     '{!r} != {!r}'.format(self.fill_value, other.fill_value)
                 msgs.append(msg)
+        if (self.cell_measures_and_dims != other.cell_measures_and_dims):
```
@marqh As with concatenate, does the order of the cell measures matter?

bjlittle commented on lib/iris/_merge.py:
```diff
@@ -1466,9 +1473,11 @@ def _get_cube(self, data):
                                for coord, dims in self._aux_coords_and_dims]
         kwargs = dict(zip(iris.cube.CubeMetadata._fields, signature.defn))
 
+        cmds = self._cell_measures_and_dims
```
@marqh Careful ... you need to copy the cell measures, otherwise multiple **different** (merged) cubes will end up sharing the **same** cell methods :bomb:

bjlittle replied:
@marqh Go on, change `cmds` -> `cm_and_dims` ... you know you want to :wink:

bjlittle commented:
@marqh Hoo-rah! :beers: There are more than 100 comments on this PR!

Hang in there, it's getting close ...

marqh commented:
I think the tests are all passing now

@bjlittle should I wait for your thoughts before I rebase?

marqh commented:
> If we have a sequence of cell *measures * for a cube and our understanding is that order doesn't matter (I'm kinda guided by your interpretation and expertise in all things CF here) then I'd be in favour of imposing a sort order which you could base on the CellMeasure._as_defn().

I think I agree.  We can then be confident of consistency in the later use of this list

bjlittle commented:
@marqh We seem to have been disconnected on the :phone: ... I assume you were buying into an imposed sort order

bjlittle commented:
@marqh Go for it :smile:

marqh commented:
i've gone for it; the tests are re-running now

if they all pass, are you content for me to rebase?

bjlittle commented:
@marqh Travis seems to be misbehaving, so I restarted the tests ...

Happy for you to rebase if they all pass.

This PR is pretty noisy :hear_no_evil: so I need to check all the comments to make sure we're addressed each point and nothing has slipped through the net.

marqh commented:
@bjlittle it seems our implementation is not python3 compatible

marqh commented:
```
Python 2.7.6 (default, Jun 22 2015, 17:58:13) 
[GCC 4.8.2] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> alist = [('a', 1), ('c', 3), ('c', 2), ('b', 2)]
>>> alist.sort(key=lambda (l, n): (l, str(n)))
>>> print(alist)
[('a', 1), ('b', 2), ('c', 2), ('c', 3)]
>>> 

```

```
Python 3.4.3 (default, Oct 14 2015, 20:28:29) 
[GCC 4.8.4] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> alist = [('a', 1), ('c', 3), ('c', 2), ('b', 2)]
>>> alist.sort(key=lambda (l, n): (l, str(n)))
  File "<stdin>", line 1
    alist.sort(key=lambda (l, n): (l, str(n)))
                          ^
SyntaxError: invalid syntax

```

marqh commented:
```
alist.sort(key=lambda atuple: (atuple[0], str(atuple[1])))
```

seems to work, slightly less readable (imo), but hey

marqh commented:
i'll leave you time to browse, commit by commit

leave me a
'go go rebase'
message if/when you are ready

many thanks
mark

bjlittle commented:
@marqh Will do, thanks!

marqh commented:
go go tests :)

rhattersley commented:
> alist.sort(key=lambda atuple: (atuple[0], str(atuple[1])))
> 
> seems to work, slightly less readable (imo), but hey

Tuple parameters were removed in Python 3 via [PEP 3113](http://legacy.python.org/dev/peps/pep-3113/). The [recommended approach](http://legacy.python.org/dev/peps/pep-3113/#transition-plan) would be to replace something like `lambda (a, b): a + b` with `lambda a_b: a_b[0] + a_b[1]`.

bjlittle commented:
@rhattersley Thanks! 

@marqh has already already used this approach :wink:

bjlittle commented on lib/iris/cube.py:
```diff
@@ -869,7 +879,7 @@ def add_aux_coord(self, coord, data_dims=None):
         self._add_unique_aux_coord(coord, data_dims)
 
-    def _add_unique_aux_coord(self, coord, data_dims):
+    def _check_multi_dim_metadata(self, metadata, data_dims, mtype):
```
@marqh You forgot to remove the `mtype` argument

bjlittle commented on lib/iris/cube.py:
```diff
@@ -879,22 +889,28 @@ def _add_unique_aux_coord(self, coord, data_dims):
 
+    def _add_unique_aux_coord(self, coord, data_dims):
+        data_dims = self._check_multi_dim_metadata(coord, data_dims,
+                                                   'coordinate')
```
@marqh Don't pass the `coordinate` string

bjlittle commented on lib/iris/cube.py:
```diff
@@ -912,6 +928,37 @@ def add_aux_factory(self, aux_factory):
+        if self.cell_measures(cell_measure):
+            raise ValueError('Duplicate cell_measures are not permitted.')
+        data_dims = self._check_multi_dim_metadata(cell_measure, data_dims,
+                                                   'cell_measure')
```
@marqh Don't pass the `cell_measure` string

bjlittle commented on lib/iris/cube.py:
```diff
@@ -912,6 +928,37 @@ def add_aux_factory(self, aux_factory):
+                                                   'cell_measure')
+        self._cell_measures_and_dims.append([cell_measure, data_dims])
+        self._cell_measures_and_dims.sort(key=lambda cmdims:
+                                          (cmdims[0]._as_defn(), cmdims[1]))
```
@marqh I like @rhattersley style `key=lambda c_d: (c_d[0]._as_defn(), c_d[1]))` ... hints to the reader with regards to unpacking that `c_d` has two elements.

bjlittle commented:
@marqh I've added a few more comments that require to be addressed.

Otherwise, functionally I'm :+1: for this PR :tada: 

I've not had time to sift through your test coverage, that's the only stone I've left unturned at this point ...

This has been quite of hot PR with an awful lot of participation from other devs @pp-mo @rhattersley @pelson @dkillick ... the time to add your two penneth is now! As I think the dust has settled on the approach of this PR ...

marqh commented:
@bjlittle 
I have made the last two changes you requested.

I have rebased to wrap up all the review changes.  I have a local copy of the unrebased branch, please let me know if you'd rather I pushed this one up instead to assist the review completion

bjlittle commented:
@marqh Awesome :+1:  Thanks for your patience!

I think it's time to merge and play :smile:
